src/runtime/threads/coroutines/detail/coroutine_impl.cpp

```diff
@@ -73,173 +69,40 @@ namespace hpx { namespace threads { namespace coroutines { namespace detail
     }
 #endif
 
-    void coroutine_impl::operator()()
+    void coroutine_impl::operator()() noexcept
     {
         typedef super_type::context_exit_status context_exit_status;
         context_exit_status status = super_type::ctx_exited_return;
 
-        // loop as long this coroutine has been rebound
-        do
+        // yield value once the thread function has finished executing
+        static result_type result_last(
+            thread_state_enum::terminated, invalid_thread_id);
+
+        std::exception_ptr tinfo;
+        try
         {
-            std::exception_ptr tinfo;
-            try
             {
-                this->check_exit_state();
-
-                HPX_ASSERT(this->count() > 0);
-
-                {
-                    coroutine_self* old_self = coroutine_self::get_self();
-                    coroutine_self self(this, old_self);
-                    reset_self_on_exit on_exit(&self, old_self);
+                coroutine_self* old_self = coroutine_self::get_self();
+                coroutine_self self(this, old_self);
+                reset_self_on_exit on_exit(&self, old_self);
 
-                    this->m_result_last = m_fun(*this->args());
+                result_last = m_fun();
 
-                    // if this thread returned 'terminated' we need to reset
-                    // the functor and the bound arguments
-                    if (this->m_result_last.first == terminated)
-                        this->reset();
-                }
-
-                // return value to other side of the fence
-                this->bind_result(&this->m_result_last);
-            }
-            catch (exit_exception const&) {
-                status = super_type::ctx_exited_exit;
-                tinfo = std::current_exception();
-                this->reset();            // reset functor
-            }
-            catch (boost::exception const&) {
-                status = super_type::ctx_exited_abnormally;
-                tinfo = std::current_exception();
                 this->reset();
```

Comment:
I'd suggest adding an assert here making sure the returned `state == terminated`.

Reply:
done.